### PR TITLE
Feature: add expiration (TTL) and dead letter (DLX) logic

### LINKED ISSUE #30: publishing to a topic exchange with a bad routing key generates an error
1- create an exchange (topic, direct, header...)
2- create a queue and bind it with a binding key
3- publish a message to the exchange, using a routing key that does not match any binding key and the exchange binding logic.

it generates this error:

TypeError: Cannot read property 'targetQueue' of undefined
    at Object.getTargetQueues .../node_modules/mock-amqplib/src/main.js:57:31) // for direct, similar error for others.

It should instead check the `mandatory` property of the message, and handle the logic described here: https://www.rabbitmq.com/publishers.html 
In the meantime a more friendly error message would help :)

### PR DESCRIPTION
**Pull request by stepanho**
See docs [here](https://www.rabbitmq.com/ttl.html) and [here](https://www.rabbitmq.com/dlx.html).
Also added [`on('return')`](https://amqp-node.github.io/amqplib/channel_api.html#channel_events) and [`alternateExchange`](https://amqp-node.github.io/amqplib/channel_api.html#channel_assertExchange) logic, so fixes #30

### CODE REVIEW

**Comment by stepanho** [comment]
I made `Channel.sendToQueue` and `Channel.publish` synchronous, see [docs](https://amqp-node.github.io/amqplib/channel_api.html#channel_publish).
On the one hand, it's **major** change, 'cause methods returned `Promise` before, but now `Boolean`.
On the other hand, it's **patch** change: using return value as Promise is mistake and should be fixed in accordance with official docs.

**Comment by stepanho** [comment]
Fixes #30

File: src/main.js [hunk 1]
```diff
@@ -115,49 +160,93 @@ const createHeadersExchange = () => {
...
   ...EventEmitter.prototype,
   close: () => {},
-  assertQueue: async queueName => {
+  assertQueue: async function (queueName, options) {
     if (!queueName) {
       queueName = generateRandomQueueName();
     }
     if (!(queueName in queues)) {
-      queues[queueName] = createQueue();
+      queues[queueName] = createQueue(options, this);
```
Comment (sempasha) [inline src/main.js]: We can't use user owned channel for managing dead letters. Managing of dead letters is RabbitMQ server responsibility, and it must not rely on channel which user can close or destroy at anytime. It is better to have direct access to exchange, specified at `x-dead-letter-exchange` header. So dead letter processing will looks like

```js
// check if message is in queue
const index = messages.indexOf(message);
if (index > -1) {
  // remove message from queue
  messages.splice(index, 1);
  
  // check if dead letter exchange specified
  const { 'x-dead-letter-exchange': exchangeName } = message.headers;
  const exchange = exchanges[exchangeName];
  
  // if dead letter exchange exists, send message to it
  if (exchange) {
    const { 'x-dead-letter-routing-key': routingKey } = message.headers;
    const deadLetter = message; // here we also may mutate message headers according to spec
    exchange.getTargetQueues(routingKey).map((queue) => queue.add(deadLetter));
  }
}
```

Comment (stepanho) [reply src/main.js]: I deattached channel from publishing/dead-lettering

**Comment by sempasha** [comment]
Hi @mkls, please, take a look this pull request.

**Comment by mkls** [comment]
Thank you, awesome contribution! I've merged it and released a new version.

**Pull request merged**